File: LeetCode/Problems/138.py
# NeetCode - Linked List
# 138. Copy List with Random Pointer
# Difficulty : Medium
# Algorithm : Linked List
# Time complexity : O(N), Space complexity : O(N)
# Runtime : 42 ms (74.35%), Memory : 17.26 MB (70.34%)

# Refernce Neetcode Solution

"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        hash_map = { None : None }  # if node.next of node.random is None -> return None

        # 1. create copy and store
        cur = head
        while cur:
            hash_map[cur] = Node(cur.val)
            cur = cur.next

        # 2. connect each copy using hash_map and original
        cur = head
        while cur:
            node = hash_map[cur]
            node.next = hash_map[cur.next]
            node.random = hash_map[cur.random]
            cur = cur.next

        # 3. return
        return hash_map[head]

File: LeetCode/Problems/138.java
/*
138. Copy List with Random Pointer
Difficulty : Medium
Algorithm : Linked List
Time complexity : O(N), Space complexity : O(1)
Runtime : 0 ms (100%), Memory : 44.14 MB (83.42%)
*/

/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

import java.util.*;

class Solution {
    public Node copyRandomList(Node head) {
        if(head == null) {return null;}
        Node node = new Node(head.val);
        Node cur = node;
        Node head_cur = head;

        Map<Node, Node> dict = new HashMap<>();
        dict.put(head_cur, cur);

        while(head_cur.next != null) {
            cur.next = new Node(head_cur.next.val);
            cur = cur.next;
            head_cur = head_cur.next;
            dict.put(head_cur, cur);
        }

        cur = node;
        head_cur = head;
        while(cur != null) {
            cur.random = dict.get(head_cur.random);
            cur = cur.next;
            head_cur = head_cur.next;
        }

        return node;
    }
}
